File: stdlib/lua/hull/kv/_handle.lua
--[[
  hull.kv._handle - the semantic surface over a backend store.

  Wraps a memory/sql store with key/value validation and capability
  enforcement (an optional op on a backend that lacks the cap throws
  `unsupported`, never silently no-ops). Shared by hull.kv and hull.cache so
  the two present the same core methods with the same guarantees; hull.cache
  layers fetch() on top. Handles are used method-style (`h:get(k)`).

  Internal module. SPDX-License-Identifier: AGPL-3.0-or-later
]]

local u = require("hull.kv._util")

local M = {}
local H = {}
H.__index = H

function M.build(store, meta)
    return setmetatable({
        _s = store, caps = store.caps,
        namespace = meta.namespace, backend = meta.backend,
    }, H)
end

local function need(self, cap)
    if not self.caps[cap] then
        u.error("unsupported",
            "kv: backend '" .. self.backend .. "' does not support " .. cap)
    end
end

function H:get(k) u.check_key(k); return self._s:get(k) end

function H:set(k, v, opts)
    u.check_key(k); u.check_value(v)
    self._s:put(k, v, opts and opts.ttl)
    return true
end

function H:delete(k) u.check_key(k); return self._s:del(k) end

function H:exists(k) u.check_key(k); return self._s:has(k) end

function H:incr(k, by, opts)
    need(self, "atomic_increment"); u.check_key(k)
    by = by or 1
    if type(by) ~= "number" or by ~= math.floor(by) then
        u.error("invalid_argument", "kv: incr amount must be an integer")
    end
    return self._s:incr(k, by, opts and opts.ttl)
end

function H:cas(k, expected, new, opts)
    need(self, "compare_exchange"); u.check_key(k); u.check_value(new)
    if expected ~= nil then u.check_value(expected) end
    return self._s:cas(k, expected, new, opts and opts.ttl)
end

function H:clear() self._s:clear(); return true end

function H:scan(prefix, opts)
    need(self, "scan")
    if prefix ~= nil and type(prefix) ~= "string" then
        u.error("invalid_argument", "kv: scan prefix must be a string (bytes)")
    end
    local limit = u.check_count(opts and opts.limit, "scan limit")
    return self._s:scan(prefix, limit)
end

function H:cleanup() return self._s:cleanup() end

function H:stats() return self._s:stats() end

return M

File: stdlib/js/hull/kv/_handle.js
/*
 * hull:kv:_handle - the semantic surface over a backend store (JS mirror of
 * hull.kv._handle). Key/value validation + capability enforcement (an optional
 * op on a backend that lacks the cap throws `unsupported`). Shared by hull.kv
 * and hull.cache; hull.cache layers fetch() on top.
 *
 * Internal module. SPDX-License-Identifier: AGPL-3.0-or-later
 */

import util from "hull:kv:_util";

function build(store, meta) {
    const need = (cap) => {
        if (!store.caps[cap])
            util.error("unsupported",
                "kv: backend '" + meta.backend + "' does not support " + cap);
    };
    return {
        namespace: meta.namespace,
        backend: meta.backend,
        caps: store.caps,

        get(k) { util.checkKey(k); return store.get(k); },

        set(k, v, opts) {
            util.checkKey(k); util.checkValue(v);
            store.put(k, v, opts && opts.ttl);
            return true;
        },

        delete(k) { util.checkKey(k); return store.del(k); },

        exists(k) { util.checkKey(k); return store.has(k); },

        incr(k, by, opts) {
            need("atomic_increment"); util.checkKey(k);
            if (by === undefined) by = 1;
            if (typeof by !== "number" || !Number.isInteger(by))
                util.error("invalid_argument", "kv: incr amount must be an integer");
            return store.incr(k, by, opts && opts.ttl);
        },

        cas(k, expected, newVal, opts) {
            need("compare_exchange"); util.checkKey(k); util.checkValue(newVal);
            if (expected !== undefined && expected !== null) util.checkValue(expected);
            return store.cas(k, expected, newVal, opts && opts.ttl);
        },

        clear() { store.clear(); return true; },

        scan(prefix, opts) {
            need("scan");
            if (prefix !== undefined && prefix !== null && typeof prefix !== "string")
                util.error("invalid_argument", "kv: scan prefix must be a string (bytes)");
            const limit = util.checkCount(opts && opts.limit, "scan limit");
            return store.scan(prefix, limit);
        },

        cleanup() { return store.cleanup(); },

        stats() { return store.stats(); },

        // Release an owned backend connection (the networked valkey/redis
        // store). A no-op for the memory/sql stores (borrowed/shared); GC
        // finalizes an unclosed network connection anyway.
        close() { if (store.close) store.close(); },

        // Exposed for hull.cache.open's fetch() (byte-cache get-or-compute).
        _store: store,
    };
}

export default { build };
